### PR TITLE
Dedicate half constraint budget to a suboptimizer

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Camille posted a problem that was not solved by Nevergrad.
With the current PR the pb is solved.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/parametrization/core.py [hunk 1]
```diff
@@ -305,6 +305,17 @@ def set_name(self: P, name: str) -> P:
...
+    def constraint_penalties(self) -> tp.List[float]:
+        """Returns the list of constraint penalties penalties
+
+        Returns
+        -------
+        list of float
+            the list of penalties for each constraint
+        """
+        val = self.value
+        return sum(utils.float_penalty(func(val)) for func in self._constraint_checkers)
```
Comment (jrapin) [inline nevergrad/parametrization/core.py:317]: isnt that a float and not a list of floats? I thing a list of floats is more general (and easy to sum, while we cannot unsum the float). Also, I expect this will soon disappear so for now we should probably keep it hidden

File: nevergrad/optimization/base.py [hunk 2]
```diff
@@ -414,7 +414,7 @@ def _update_archive_and_bests(self, candidate: p.Parameter, loss: tp.FloatLoss)
         if self.pruning is not None:
             self.archive = self.pruning(self.archive)
 
-    def ask(self) -> p.Parameter:
+    def ask(self, ignore_constraints: bool = False) -> p.Parameter:
```
Comment (jrapin) [inline nevergrad/optimization/base.py:417]: please don't change the ask and tell API

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/parametrization/core.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/parametrization/core.py:317]: Right! Fixed.

File: nevergrad/optimization/base.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/optimization/base.py]: Fixed.

File: nevergrad/optimization/base.py [hunk 3]
```diff
@@ -736,3 +749,25 @@ def __eq__(self, other: tp.Any) -> tp.Any:
...
                 return True
         return False
+
+
+def _constraint_solver(parameter: p.Parameter, budget: int) -> p.Parameter:
+    """Runs a suboptimization to solve the parameter constraints"""
+    parameter_without_constraint = parameter.copy()
+    parameter_without_constraint._constraint_checkers.clear()
+    opt = registry["OnePlusOne"](parameter_without_constraint, num_workers=1, budget=budget)
+    opt._constraints_manager.max_trials = 1
```
Comment (jrapin) [inline nevergrad/optimization/base.py:759]:
```suggestion
```

not needed anymore I guess

File: nevergrad/parametrization/core.py [hunk 4]
```diff
@@ -241,6 +241,17 @@ def __repr__(self) -> str:
 
     # %% Constraint management
 
+    def _constraint_penalties(self) -> float:
+        """Returns the list of constraint penalties penalties
```
Comment (jrapin) [inline nevergrad/parametrization/core.py:245]: please update the docstring, and probably the name if you do not want to return the list of penalities.
I still think the list of penalties is better since it provides more flexibility though (eg: multiobjective)

Actually, returning a list of floats would be much better since this be used in the `satisfies_constraints` method just below, and avoid duplication, so it's probably the way to go, and you can just sum in the `_constraint_solver` function.

File: nevergrad/optimization/base.py [hunk 5]
```diff
@@ -736,3 +749,25 @@ def __eq__(self, other: tp.Any) -> tp.Any:
...
+    for _ in range(budget):
+        cand = opt.ask()
+        # Our objective function is minimum for the point the closest to
+        # the original candidate under the constraints.
+        penalty = cand._constraint_penalties()
+        if not penalty > 0:  # constraints are satisfied
+            return cand
```
Comment (jrapin) [inline nevergrad/optimization/base.py:766]: don't return this, since this candidate does not have any constraint anymore listed, it could create weird bugs.

File: nevergrad/optimization/base.py [hunk 6]
```diff
@@ -736,3 +749,25 @@ def __eq__(self, other: tp.Any) -> tp.Any:
...
+    """Runs a suboptimization to solve the parameter constraints"""
+    parameter_without_constraint = parameter.copy()
+    parameter_without_constraint._constraint_checkers.clear()
+    opt = registry["OnePlusOne"](parameter_without_constraint, num_workers=1, budget=budget)
+    opt._constraints_manager.max_trials = 1
+    for _ in range(budget):
+        cand = opt.ask()
+        # Our objective function is minimum for the point the closest to
+        # the original candidate under the constraints.
+        penalty = cand._constraint_penalties()
```
Comment (jrapin) [inline nevergrad/optimization/base.py:764]: isnt there no more constraint? I don't understand how this could work, you probably need to get the state, set the state to a parameter with the loss, and then check for penalty

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/optimization/base.py (hunk 6 above)
Comment (jrapin) [reply nevergrad/optimization/base.py:764]: Actually simper:
```python
penalty = sum(utils.float_penalty(func(cand.value)) for func in parameter._constraint_checkers)
```

And we can just delete `_constraint_penalties`

File: nevergrad/parametrization/core.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/parametrization/core.py:245]: Deleted as you suggested.

File: nevergrad/optimization/base.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/optimization/base.py:766]: oh my god thx.

File: nevergrad/optimization/base.py (hunk 6 above)
Comment (teytaud) [reply nevergrad/optimization/base.py:764]: Thx!

File: nevergrad/optimization/base.py [hunk 7]
```diff
@@ -736,3 +749,25 @@ def __eq__(self, other: tp.Any) -> tp.Any:
...
+        # Our objective function is minimum for the point the closest to
+        # the original candidate under the constraints.
+        penalty = sum(utils._float_penalty(func(cand.value)) for func in parameter._constraint_checkers)
+
+        # TODO: this may not scale well with dimension
+        distance = np.tanh(np.sum(cand.get_standardized_data(reference=parameter) ** 2))
+        # TODO: because of the return whenever constraints are satisfied, the first case never arises
+        loss = distance if penalty <= 0 else penalty + distance + 1.0
+        opt.tell(cand, loss)
+        if not penalty > 0:  # constraints are satisfied
```
Comment (jrapin) [inline nevergrad/optimization/base.py:770]:
```suggestion
        if penalty <= 0:  # constraints are satisfied
```

**Review by jrapin — APPROVED** [review]

File: nevergrad/optimization/base.py [hunk 8]
```diff
@@ -433,19 +433,29 @@ def ask(self) -> p.Parameter:
         current_num_ask = self.num_ask
         # tentatives if a cheap constraint is available
         # TODO: this should be replaced by an optimization algorithm.
-        max_trials = self._constraints_manager.max_trials
+        max_trials = max(
+            1,
+            self._constraints_manager.max_trials // 2
+            if self.budget is None
+            else (self.budget - self.num_ask),
```
Comment (jrapin) [inline nevergrad/optimization/base.py:440]: why `self.budget -  self.num_ask`? right now this does not make sense to me, and this seems unexpected

**Review by jrapin — CHANGES_REQUESTED** [review]
Actually, requesting changes again to make sure this is not merge before I understand the budget

File: nevergrad/optimization/base.py (hunk 8 above)
Comment (teytaud) [reply nevergrad/optimization/base.py:440]: If the optimizer has a maximum budget, we want the budget we use to be smaller than the remaining budget before the optimizer crashes, hence self.budget - self.num_ask.

Comment (jrapin) [reply nevergrad/optimization/base.py:440]: so you probably want to do some kind of mean between self._constraints_manager.max_trials and `self.budget - self.num_ask` so that:
you don't go beyond max_trials, and you dont exhaust more than half the remaining budget of the optimizer

Comment (teytaud) [reply nevergrad/optimization/base.py:440]: min(self._constraints_manager.max_trials, self.budget - self.num_ask)   could make sense.
Eventually all that stuff is going to be equipped with parameters for combining the auxiliary solver and the simple rejection of mutations.
Besides adding the famous constraint learning tool for expensive constraints that would be useful for some of our users... still a long way to go on the path to constraint management :-)

**Review by jrapin — APPROVED** [review]

**Comment by teytaud** [comment]
Thx @jrapin , you deserve more credit than me for this PR.
